### ISSUE TITLE
DeprecationWarning: the MimeWriter module is deprecated; use the email package instead

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4038]
|<img alt="maurice's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| maurice reported|
|-|-|
|Trac ID|trac#4038|
|Type|defect|
|Created|2009-09-24 12:54:41Z|
|Branch|https://github.com/twisted/twisted/tree/fix-mimewriter-deprecation-warning-4038|

In python 2.6.2 the following DeprecationWarning is given on:

```
>>> from twisted.mail import smtp
/usr/lib/python2.6/dist-packages/twisted/mail/smtp.py:10: DeprecationWarning: the MimeWriter module is deprecated; use the email package instead
  import MimeWriter, tempfile, rfc822
```

In smpty.py the MimeWriter module is imported.
MimeWriter is used in sendEmail.

Even if sendmail is used in stead of sendEmail, this message is given.
It is currently not possible for Twisted users to work around this warning.

<details><summary>Searchable metadata</summary>

```
trac-id__4038 4038
type__defect defect
reporter__maurice maurice
priority__high high
milestone__Python_2_6 Python-2.6
branch__branches_fix_mimewriter_deprecation_warning_4038 branches/fix-mimewriter-deprecation-warning-4038
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__easy easy
time__1253796881000000 1253796881000000
changetime__1277919097000000 1277919097000000
version__None None
owner__ 
cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

`sendEmail` has been deprecated since mid-2003.  We can probably just delete it now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Ooh.  Look, twisted/news/database.py uses `smtp.sendEmail`.  Blech.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

[#4381](https://github.com/twisted/twisted/issues/4381) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ChrisW's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|ChrisW commented|
|-|-|

A workaround is to add a warning filter before you import the defective module:

import warnings
warnings.filterwarnings(
    'ignore','^the MimeWriter',DeprecationWarning,'twisted'
    )

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

[#4484](https://github.com/twisted/twisted/issues/4484) was a duplicate.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @cyli<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_12"></a>|@cyli commented|
|-|-|

(In [29462]) Branching to 'fix-mimewriter-deprecation-warning-4038'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_13"></a>|@cyli commented|
|-|-|

(In [29463]) Removed some unused imports from smtp, and made the MimeWriter import local so that the horrible errors will only occur when the already-deprecated sendEmail is used

refs [#4038](https://github.com/twisted/twisted/issues/4038)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli removed owner|
|-|-|

Also, I opened [#4528](https://github.com/twisted/twisted/issues/4528) and [#4529](https://github.com/twisted/twisted/issues/4529) in hopes of eventually being able to remove sendEmail.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @cyli|
|-|-|

Thanks.  It's nice to make progress against the heap of warnings spewed out instantly on starting `trial` and many other Twisted apps.

Just one comment.  I'm a little uncomfortable with the removals of `twisted.cred.checkers` and `twisted.cred.credentials` imports while code in the module still uses expressions like `cred.credentials.Anonymous` and `cred.error.UnauthorizedLogin`.  Can you leave the imports alone or replace them and these expressions with something a little saner?

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_16"></a>|@cyli commented|
|-|-|

(In [29486]) Put back the twisted.cred.credentials and twisted.cred.checkers imports

refs [#4038](https://github.com/twisted/twisted/issues/4038)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @cyli|
|-|-|

Great.  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_19"></a>|@cyli set status to `closed`|
|-|-|

(In [29521]) Merge fix-mimewriter-deprecation-warning-4038 : importing t.mail.smtp no longer gives MimeWriter deprecation warning

Author: cyli
Reviewer: exarkun
Fixes: [#4038](https://github.com/twisted/twisted/issues/4038)

Previously importing t.mail.smtp caused a deprecation warning about MimeWriter (deprecated since Python 2.3) to be emitted.  Now only using smtp.sendEmail (deprecated since mid-2003) emits that warning.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
